design-proposals/klip-4-custom-types.md

```diff
@@ -0,0 +1,186 @@
+
+Kafka often contains very complex nested data structures and it is common to see schemas nested with
+other schemas. It is desirable to be able to define this custom schema in KSQL once and reference it
+in all future statements. For example, imagine the following schema declarations:
+
+```sql
+CREATE TYPE ADDRESS AS STRUCT<number INTEGER, street VARCHAR, city VARCHAR>;
+CREATE TYPE PERSON AS STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS>;
+CREATE TYPE COMPANY AS STRUCT<name VARCHAR, headquarters ADDRESS>;
+```
+
+With this feature organizations can leverage existing schemas in their KSQL queries and share 
+data with other parts of their organizations.
+
+## Scope
+
+* Syntax to register and unregister custom type aliases
+* Ability to include custom types at compile-time (e.g. in the JAR) for use in UDF declarations
+* Ability to list all custom types
+
+Schema Registry integration is not considered within scope.
+
+## Value/Return
+
+Real data is complex, nested and referential. This feature will help us dramatically improve the
+user interaction with complicated schemas. 
+
+## Public APIS
+
+### CREATE TYPE
+
+```sql
+CREATE TYPE <type_name> AS <type>;
+```
+
+The `CREATE TYPE` syntax will allow KSQL users to register a type alias directly in SQL (either 
+interactive or headless modes). Any types registered using this command can be leveraged in any
+future statement. 
+
+Any attempts to register the same type twice without a corresponding `DROP TYPE` will fail.
+
+### DROP TYPE 
+
+```sql
+DROP TYPE <type_name>;
+```
+
+The `DROP TYPE` syntax will allow KSQL users to remove a type alias from KSQL. This statement
+will fail if the type is being used in any active query or UDF.
+
+### SHOW TYPES
+
+```sql
+SHOW TYPES;
+```
+
+The `SHOW TYPES` command will list all custom types and their type definitions. A sample output
+for this command would be:
+```
+|---------------|--------------------------------------------------------------|------------------|
+|     name      |       definition                                             | source           |
+|---------------|--------------------------------------------------------------|------------------|
+| ADDRESS       | STRUCT<number INTEGER, street VARCHAR, city VARCHAR>         | types.sql        |
+| PERSON        | STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS> | types.sql        |
+| COMPANY       | STRUCT<name VARCHAR, headquarters ADDRESS>                   | CLI              |
+|---------------|--------------------------------------------------------------|------------------|
+```
+
+### Extension Directory
+
+The `ksql.extension.dir` will now also recognize `.sql` files that contain only `CREATE TYPE`
+commands. These commands will all be run before compiling any UDFs so these custom type declarations
+can be used in the `@Udf` annotations. They will be loaded in natural order to ensure that the
+loading behavior is deterministic.
```

Comment:
I can imagine this would result in many mistakes by users. But the way we have implemented the UDFs makes it difficult to have any other option here!